xwiki-platform-core/xwiki-platform-eventstream/xwiki-platform-eventstream-api/src/main/java/org/xwiki/eventstream/internal/UntypedEventListener.java

```diff
@@ -0,0 +1,190 @@
+
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.component.manager.ComponentLookupException;
+import org.xwiki.component.manager.ComponentManager;
+import org.xwiki.eventstream.UntypedRecordableEvent;
+import org.xwiki.eventstream.UntypedRecordableEventDescriptor;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.observation.AbstractEventListener;
+import org.xwiki.observation.ObservationManager;
+import org.xwiki.observation.event.AllEvent;
+import org.xwiki.observation.event.Event;
+import org.xwiki.rendering.block.XDOM;
+import org.xwiki.rendering.renderer.BlockRenderer;
+import org.xwiki.rendering.renderer.printer.DefaultWikiPrinter;
+import org.xwiki.rendering.renderer.printer.WikiPrinter;
+import org.xwiki.script.ScriptContextManager;
+import org.xwiki.template.Template;
+import org.xwiki.template.TemplateManager;
+
+/**
+ * This listener listens to every event occurring on the wiki. When an event happens, it tries to determine if this
+ * event should trigger another one (an {@link UntypedRecordableEvent} which is described by a
+ * {@link UntypedRecordableEventDescriptor}) given the current context.
+ *
+ * @version $Id$
+ * @since 9.5RC1
+ */
+@Component
+@Singleton
+@Named(UntypedEventListener.NAME)
+public class UntypedEventListener extends AbstractEventListener
+{
+    /**
+     * The listener name.
+     */
+    public static final String NAME = "Untyped Event Listener";
+
+    /**
+     * The binding name of the event when a validation expression is rendered.
+     */
+    public static final String EVENT_BINDING_NAME = "event";
+
+    @Inject
+    private ObservationManager observationManager;
+
+    @Inject
+    private TemplateManager templateManager;
+
+    @Inject
+    @Named("plain/1.0")
+    private BlockRenderer plainRenderer;
+
+    @Inject
+    @Named("wiki")
+    private ComponentManager componentManager;
+
+    @Inject
+    private ModelBridge modelBridge;
+
+    @Inject
+    private ScriptContextManager scriptContextManager;
+
+    @Inject
+    private Logger logger;
+
+    /**
+     * Constructs a {@link UntypedEventListener}.
+     */
+    public UntypedEventListener()
+    {
+        super(NAME, AllEvent.ALLEVENT);
+    }
+
+    @Override
+    public void onEvent(Event event, Object source, Object data)
+    {
+        try {
+            // Get every UntypedEventDescriptor registered in the ComponentManager
+            List<UntypedRecordableEventDescriptor> descriptors =
+                    componentManager.getInstanceList(UntypedRecordableEventDescriptor.class);
+
+            // Filter the event descriptors concerned by the event, then create the concerned events
+            for (UntypedRecordableEventDescriptor descriptor : descriptors) {
+                // If the event is expected by our descriptor
+                if (descriptor.getEventTriggers().contains(event.getClass().getCanonicalName())
```

Comment:
I don’t think that I get your point. 

Presently, the event type that is stored in an `UntypedRecordableEvent` acts as a sort of "unique identifier" ; it could be something like `BlogPostPublished`, or `ForumThreadAdded` … it is a way to identify the type of event we are dealing with without relying on its class name. 

The event type of an `UntypedRecordableEventDescriptor` is the same unique identifier. Therefore, with an `UntypedEvent`, we can get the corresponding descriptor with `componentManager.getInstance(UntypedRecordableEventDescriptor.class, event.getEventType())`.

With that in mind the event triggers represent a list of `Event` (not `RecordableEvent`) that should trigger the particular `UntypedEvent`. This list could be formed of `com.xpn.xwiki.internal.event.XObjectAddedEvent` or `org.xwiki.bridge.event.DocumentCreatedEvent` for example.

So, if we use the event triggers list as a hint for the `RecordableEventDescriptor`, we will have to join it, as hints are AFAIK only String-based, it will not be very clean and will involve some computation time in order to split the String and convert it back to a List every time an event is raised.

On the other side we do have to get all the descriptors all the time, as multiple descriptors can be triggered by the same event.

Sorry for the long post …